w3-deal.md

```diff
@@ -0,0 +1,217 @@
+}
+
+// Note that we use capital case field names for compatibility with go
+export interface DealProposal {
+  Piece: Piece.Link
+  Size: Piece.PaddedSize
+
+  VerifiedDeal: boolean
+
+  Client: Address
+  Provider: Address
+  Label: DealLabel
+
+  StartEpoch: ChainEpoch
+  EndEpoch: ChainEpoch
+  StoragePricePerEpoch: TokenAmount
+
+  ProviderCollateral: TokenAmount
+  ClientCollateral: TokenAmount
+}
+
+export interface Address {
+  addr: string
+}
+
+export interface DealLabel {
+ bs: Uint8Array
+ notString: boolean
+}
+
+export type ChainEpoch = Uint68
+export type TokenAmount = ToString<bigint>
+
+export type Signature = Uint8Array
+
+export type SignError =
+  | InvocationError
+  | InvalidDeal // ProposalRemarshalMismatch
+  | SigningError // WalletSignError
+```
+
+## Capabilities
+
+### `deal/sign`
+
+Broker can invoke `deal/sign` capability with `DealProposal` in (`nb` field). Storefront MUST encode supplied `DealProposal` (`nb` field) into a CBOR block and then sign it with a wallet private key.
+
+Provider MUST response with raw bytes of the signature.
+
+## HTTP Interface
+
+Given that some actors (e.g. Spade) do not support UCANs natively they are not able to send signed invocations. As compromise protocol implementer is RECOMMENDED to expose plain HTTP API that trade-offs some security for convenience of interop.
+
+Implementation MUST expose HTTP POST endpoint that accepts `application/vnd.ucan.cbor` requests with CBOR encoded as payload. These requests MUST provide `Authorization: Bearer` header with a UCAN delegation authorizing a request. Receiving principal MUST derive invocation from the provided `Authorization` and set invoked capability `nb` field to decoded CBOR block of the payload. Receiving principal MUST execute received capability and encode result of the invocation as an HTTP response.
+
+> Above HTTP interface could be utilize by spade to obtain signatures from w3up without having to proxy them through UCAN proxy.
+
+## Interaction Flow
+
+### Authorization per aggregate
+
+_Storefront_ (w3up) MAY delegate `deal/sign` UCAN capability to the _Agency_ (spade-proxy) and specify `Piece` and `Size` fields of the submitted aggregate.
+
+_Agency_ (spade-proxy) could also re-delegate that capability to the _Agency_ (spade) allowing it to request signature directly from _Storefront_ front.
+
+> ℹ️ Since ♠️ Spade does not support UCANs (yet), _Agency_ could instead create an invocation UCAN and pass it on to Spade so it could be used by spade as a plain, but short lived, JWT token for signing that specific deal.
+
+### Long term authorization
+
+_Storefront_ (w3up) MAY delegate unconstrained `deal/sign` UCAN capability to the _Agency_ (spade-proxy). By leaving out `Piece` and `Size` fields it will authorize it to sign any deals.
+
+_Agency_ could also re-delegate same unconstrained `deal/sign` UCAN capability to the _Agency_ (spade-proxy) allowing it to sign any deals.
+
+This trade-offs increased security for convenience.
```

Comment:
We can delegate capability when we perform `aggregate/offer` individually per piece (aggregate). When `spade-proxy` receives the offer it can re-delegate to the broker. Given we can easily have this, should we consider to have it required instead of the long term solution?

Reply:
I think we should go for auth per offer, that said I think it's good to have both options listed. There is another tradeoff that I failed to communicate here, which is token per offer means that those tokens are likely to be included in requests and consequently can get leaked very easily. Long term tokens are likely exchanged ahead of time and out of bound, so they are less likely to get leaked in comparison to per-request approach.